### PR TITLE
docs: Add high-level documentation of abstractions

### PR DESCRIPTION
**Pull request by jjerphan**
#### Reference Issues/PRs

None.

#### What does this implement/fix? Explain your changes.

Provide some high level general documentation of concepts.

#### Any other comments?

This still needs to be completed.

### CODE REVIEW

File: cpp/arcticdb/storage/storages.hpp [hunk 1]
```diff
@@ -21,6 +21,15 @@
...
 namespace arcticdb::storage {
 
+/*
+ * The Storages class abstracts over multiple physical stores and controls how ArcticDB read and write to multiple
+ * ones.
+ *
+ * Reasons for Storages to exist:
+ *  - Disaster recovery for Storages: if the first storage fails we can fall back to the second one, etc.
+ *  - Tiered storage: recent data goes to a fast, comparatively expensive storage and then is gradually moved
+ *  into a slower, cheaper one.
```
Comment (alexowens90) [inline cpp/arcticdb/storage/storages.hpp:31]: These are the long term aims, its worth noting that neither feature is implemented right now

Comment (jjerphan) [reply cpp/arcticdb/storage/storages.hpp:28]: Addressed by ccf5e71.

File: cpp/arcticdb/async/async_store.hpp [hunk 2]
```diff
@@ -19,6 +19,10 @@
 
 namespace arcticdb::async {
 
+/*
+ * AsyncStore is a wrapper around a Store that provides async methods for writing and reading data.
+ * It is used by the VersionStore to write data to the Store in a non-blocking way.
+ */
```
[the remark this answers is not in the document]
Comment (jjerphan) [reply cpp/arcticdb/async/async_store.hpp:32]: Addressed by ccf5e71.

File: cpp/arcticdb/column_store/chunked_buffer.hpp [hunk 3]
```diff
@@ -26,6 +26,14 @@
 
 namespace arcticdb {
 
+/*
+ * ChunkedBufferImpl is an untyped buffer that is composed of blocks of data that can be either regularly or
+ * irregularly sized, with optimizations for both cases as well as the case where there is a single contiguous
```
Comment (alexowens90) [inline cpp/arcticdb/column_store/chunked_buffer.hpp:31]: There are optimizations for:

- A single block
- Multiple blocks that are all regularly sized
- Multiple blocks that are regularly sized up until a point, and irregularly sized after that point.

If the blocks are all irregularly sized there's nothing we can do to optimize.

File: cpp/arcticdb/entity/native_tensor.hpp [hunk 4]
```diff
@@ -17,6 +17,12 @@ inline ssize_t calc_elements(const shape_t* shape, ssize_t ndim) {
     return std::accumulate(shape, shape + ndim, ssize_t(1), std::multiplies<ssize_t>());
 }
 
+/*
+ * A wrapper around a 2D tensor that provides a more convenient interface for accessing the data
```
Comment (alexowens90) [inline cpp/arcticdb/entity/native_tensor.hpp:21]: 1D or 2D

File: cpp/arcticdb/storage/test/in_memory_store.hpp [hunk 5]
```diff
@@ -20,6 +20,7 @@
 #include <vector>
 
 namespace arcticdb {
+    /* In-memory implementation of Store for testing purposes. */
```
Comment (alexowens90) [inline cpp/arcticdb/storage/test/in_memory_store.hpp:23]: Probably worth noting that a lot of methods are not implemented for this store (although @joe-iddon might be changing that soon)

**Review by alexowens90 — CHANGES_REQUESTED** [review]

**Review by jamesmunro — APPROVED** [review]

**Review by alexowens90 — APPROVED** [review]

**Pull request merged**